src/Bicep.Core/Analyzers/Linter/DependsOnRemovalRewriter.cs

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using Bicep.Core.Visitors;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Bicep.Core.Analyzers.Linter
+{
+    // Looks for resources where a dependency can already be inferred by the structure of the resource declaration.
+    // 
+    // As an example, because the below resource already has a reference to 'otherRes' in the name property, the dependsOn is not adding anything:
+    //   resource myRes 'My.Rp/myResource@2020-01-01' = {
+    //     name: otherRes.name
+    //     dependsOn: [
+    //       otherRes
+    //     ]
+    //   }
+    public class DependsOnRemovalRewriter : SyntaxRewriteVisitor
+    {
+        private readonly SemanticModel semanticModel;
+
+        public DependsOnRemovalRewriter(SemanticModel semanticModel)
+        {
+            this.semanticModel = semanticModel;
+        }
+
+        private SyntaxBase? TryGetReplacementValue(SyntaxBase value) =>
+            value switch
+            {
+                ObjectSyntax @object => TryGetReplacementBody(@object),
+                IfConditionSyntax ifCondition => TryGetReplacementIfCondition(ifCondition),
+                _ => null
+            };
```

Comment:
Does this handle for-expressions correctly?

Reply:
The rule using this has been temporarily removed, so this can wait.

Reply:
https://github.com/Azure/bicep/issues/2776